### DISCUSSION TITLE
[24 Gennaio 2023] Pipeline

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#79] [Pipeline (M-Z)]
Il programma qui di seguito somma i primi L numeri nella sequenzafornita, limitando il valore massimo di ogni elemento a 4 (un elemento maggiore di 4 verrà limitato al valore 4). La stringa finale viene stampata se la somma totale è maggiore o uguale a 12.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
$1)$ le istruzioni tra le quali sono presenti **data hazard** – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
$2)$ le istruzioni tra le quali sono presenti **control hazard** – indicare i numeri di istruzione N ed M (nel formato N / M);
$3)$ quanti cicli di clock sono necessari ad eseguire il programma tramite **forwarding**, spiegando il calcolo effettuato;
$4)$ quanti cicli di clock sarebbero necessari ad eseguire il programma **senza forwarding**, spiegando il calcolo effettuato;
$5)$ quali sono, per ognuna delle cinque fasi, le **istruzioni** (o le bolle) in pipeline durante **il 13° ciclo di clock** (con **forwarding**);
$6)$ qual è il ruolo delle connessioni al multiplexer a monte del PC nel diagramma seguente, spiegando nel maggior dettaglio possibile i tre possibili input e il ruolo dei due bit di controllo.
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/5a646ba3-23ac-4668-8198-834ec2b3872e)
![Architettura Pipeline](https://github.com/Elia-Belli/ae2223/assets/107944476/3f159f17-c867-4d6e-af09-d2ecf117a903)

---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11666023/AE.2021-22.-.Appello.4.-.2023-01-24.-.Esercizio.pipeline.-.Scheda.di.calcolo.pdf)

**Nota (@matypist, 17/01/2025):**
- Nella soluzione del professore al punto **5** in realtà viene considerato il **14°** ciclo di clock (anziché il **13°** richiesto dalla consegna)

**Comment by Frisbon** [comment]
Ringrazio **J** onut Cicio per il suo metodo letteralmente OP.
_@CuriousCI_

_Soluzione combacia con quella del prof :)_

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/84731341/790800a8-cba7-4abd-ac73-548e9b3c83bf)
